File: ui/src/types/runs.ts
/**
 * Run contract types — Electron / React side
 *
 * These are the exact TypeScript mirrors of run_state.py and run_controller.py.
 * Keep in sync with the Python definitions. The FastAPI /runs endpoints
 * return these shapes; the SSE stream emits RunEvent payloads.
 *
 * Place at: ui/src/types/runs.ts
 */

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

export type RunStatus =
  | "queued"
  | "starting"
  | "running"
  | "paused"
  | "stopping"
  | "completed"
  | "failed"
  | "cancelled";

export type RunStage =
  | "initialising"
  | "navigating"
  | "handling_consent"
  | "extracting"
  | "normalising"
  | "paginating"
  | "retrying"
  | "checkpointing"
  | "storing"
  | "finishing";

export type FailureSeverityLevel = "low" | "medium" | "high" | "critical";

export type RunEventType =
  | "state_changed"
  | "progress"
  | "record_extracted"
  | "warning"
  | "error"
  | "log"
  | "checkpoint_saved"
  | "run_completed"
  | "run_failed"
  | "run_cancelled";

// ---------------------------------------------------------------------------
// Structured problem records
// ---------------------------------------------------------------------------

export interface RunWarning {
  code: string;
  message: string;
  stage: RunStage;
  timestamp: number;
  context: Record<string, unknown>;
}

export interface RunError {
  code: string;
  message: string;
  stage: RunStage;
  severity: FailureSeverityLevel;
  retryable: boolean;
  timestamp: number;
  context: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// RunState — the full state object returned by GET /runs/{run_id}
// ---------------------------------------------------------------------------

export interface RunState {
  // Identity
  run_id: string;
  scraper_id: string;
  scraper_name: string;

  // Lifecycle
  status: RunStatus;
  stage: RunStage;

  // Progress
  progress: number;            // 0.0 → 1.0
  records_extracted: number;
  records_stored: number;
  pages_visited: number;
  pages_total: number | null;

  // Health
  warnings: RunWarning[];
  errors: RunError[];
  warning_count: number;
  error_count: number;
  retry_count: number;
  retry_max: number;

  // Timing
  queued_at: number;
  started_at: number | null;
  updated_at: number;
  completed_at: number | null;
  duration_seconds: number | null;

  // Config snapshot
  target_url: string;
  extraction_mode: "dom" | "api" | "network" | "hybrid";
  proxy_enabled: boolean;
  stealth_enabled: boolean;
  scheduled: boolean;
  schedule_expr: string | null;
  resumable: boolean;
  checkpoint_path: string | null;
}

// Lightweight version for list views
export type RunSummary = Pick<
  RunState,
  | "run_id"
  | "scraper_id"
  | "scraper_name"
  | "status"
  | "stage"
  | "progress"
  | "records_extracted"
  | "warning_count"
  | "error_count"
  | "duration_seconds"
  | "started_at"
  | "queued_at"
>;

// ---------------------------------------------------------------------------
// SSE Event payloads
// ---------------------------------------------------------------------------

export interface RunEvent<T = Record<string, unknown>> {
  run_id: string;
  timestamp: number;
  data: T;
}

export type RunEventMap = {
  state_changed:      RunState;
  progress:           { progress: number; pages_visited: number; pages_total: number | null };
  record_extracted:   { records_extracted: number; record: unknown };
  warning:            { code: string; message: string };
  error:              { code: string; message: string; severity: FailureSeverityLevel; retryable: boolean };
  log:                { level: "debug" | "info" | "warning" | "error"; message: string };
  checkpoint_saved:   { checkpoint_path: string };
  run_completed:      RunState;
  run_failed:         RunState;
  run_cancelled:      RunState;
};

// ---------------------------------------------------------------------------
// API request types
// ---------------------------------------------------------------------------

export interface StartRunRequest {
  scraper_id: string;
  scraper_name: string;
  target_url: string;
  extraction_mode?: "dom" | "api" | "network" | "hybrid";
  proxy_enabled?: boolean;
  stealth_enabled?: boolean;
  schedule_expr?: string;
  priority?: number;
  resume_from?: string;
}

// ---------------------------------------------------------------------------
// UI helpers
// ---------------------------------------------------------------------------

/** Map RunStatus to a display colour token for the Runs UI */
export const STATUS_COLOR: Record<RunStatus, string> = {
  queued:    "#888780",   // gray
  starting:  "#EF9F27",   // amber
  running:   "#1D9E75",   // teal
  paused:    "#378ADD",   // blue
  stopping:  "#EF9F27",   // amber
  completed: "#639922",   // green
  failed:    "#E24B4A",   // red
  cancelled: "#888780",   // gray
};

/** Human-readable stage labels for the live progress indicator */
export const STAGE_LABEL: Record<RunStage, string> = {
  initialising:     "Starting up...",
  navigating:       "Navigating to page",
  handling_consent: "Handling consent",
  extracting:       "Extracting data",
  normalising:      "Normalising records",
  paginating:       "Moving to next page",
  retrying:         "Retrying after failure",
  checkpointing:    "Saving checkpoint",
  storing:          "Writing to storage",
  finishing:        "Finishing up",
};

/** Returns true if a run can be paused from the UI */
export const canPause = (status: RunStatus): boolean =>
  status === "running";

/** Returns true if a run can be resumed from the UI */
export const canResume = (status: RunStatus): boolean =>
  status === "paused";

/** Returns true if a run can be stopped from the UI */
export const canStop = (status: RunStatus): boolean =>
  ["queued", "starting", "running", "paused"].includes(status);

/** Returns true if the run is in a terminal state */
export const isTerminal = (status: RunStatus): boolean =>
  ["completed", "failed", "cancelled"].includes(status);

File: src/api/routers/runs.py
"""
FastAPI router — /runs

Exposes RunController to the Electron UI over HTTP + SSE.
Drop this file into src/api/routers/runs.py and register it in src/api/main.py.

Follows the exact same style as the existing routers in:
  - src/api/routers/failures.py
  - src/api/routers/feature_flags.py
"""

from __future__ import annotations

from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field

from .run_state import RunStatus
from .run_controller import RunController

router = APIRouter(prefix="/runs", tags=["runs"])


# ---------------------------------------------------------------------------
# Dependency — singleton controller injected by FastAPI
# ---------------------------------------------------------------------------

def get_controller() -> RunController:
    """
    Override this in main.py with the actual singleton:

        from .run_controller import RunController
        _controller = RunController()

        @app.on_event("startup")
        async def _startup():
            await _controller.start()

        app.dependency_overrides[get_controller] = lambda: _controller
    """
    raise RuntimeError("RunController not initialised — check lifespan setup")


# ---------------------------------------------------------------------------
# Request / Response schemas — Pydantic v2, matching existing schemas.py style
# ---------------------------------------------------------------------------

class StartRunRequest(BaseModel):
    scraper_id: str = Field(..., min_length=1)
    scraper_name: str = Field(..., min_length=1)
    target_url: str = Field(..., min_length=1)
    extraction_mode: str = Field("dom", pattern="^(dom|api|network|hybrid)$")
    proxy_enabled: bool = False
    stealth_enabled: bool = True
    schedule_expr: Optional[str] = None
    priority: int = Field(0, ge=0, le=10)
    resume_from: Optional[str] = None


class RunSummary(BaseModel):
    """Lightweight representation for list endpoints."""
    run_id: str
    scraper_id: str
    scraper_name: str
    status: str
    stage: str
    progress: float
    records_extracted: int
    warning_count: int
    error_count: int
    duration_seconds: Optional[float]
    started_at: Optional[float]
    queued_at: float


class RunDetail(RunSummary):
    """Full run state for the detail / stream views."""
    target_url: str
    extraction_mode: str
    proxy_enabled: bool
    stealth_enabled: bool
    scheduled: bool
    schedule_expr: Optional[str]
    resumable: bool
    checkpoint_path: Optional[str]
    retry_count: int
    retry_max: int
    records_stored: int
    pages_visited: int
    pages_total: Optional[int]
    warnings: list
    errors: list
    updated_at: float
    completed_at: Optional[float]


# ---------------------------------------------------------------------------
# Endpoints
# ---------------------------------------------------------------------------

@router.post("", response_model=RunDetail, status_code=202)
async def start_run(
    body: StartRunRequest,
    controller: RunController = Depends(get_controller),
):
    """
    Enqueue a new scraping run.
    Returns immediately with status=queued; poll /runs/{run_id} or
    subscribe to /runs/{run_id}/stream for live updates.
    """
    state = await controller.start_run(
        scraper_id=body.scraper_id,
        scraper_name=body.scraper_name,
        target_url=body.target_url,
        extraction_mode=body.extraction_mode,
        proxy_enabled=body.proxy_enabled,
        stealth_enabled=body.stealth_enabled,
        schedule_expr=body.schedule_expr,
        priority=body.priority,
        resume_from=body.resume_from,
    )
    return RunDetail(**state.to_dict())


@router.get("", response_model=List[RunSummary])
async def list_runs(
    status: Optional[str] = Query(None),
    scraper_id: Optional[str] = Query(None),
    limit: int = Query(50, ge=1, le=200),
    controller: RunController = Depends(get_controller),
):
    """Return runs filtered by status and/or scraper, newest first."""
    status_enum = RunStatus(status) if status else None
    runs = controller.list_runs(status=status_enum, scraper_id=scraper_id, limit=limit)
    return [RunSummary(**r.to_dict()) for r in runs]


@router.get("/{run_id}", response_model=RunDetail)
async def get_run(
    run_id: str,
    controller: RunController = Depends(get_controller),
):
    """Full state snapshot for a single run."""
    try:
        state = controller.get_run(run_id)
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Run not found: {run_id}")
    return RunDetail(**state.to_dict())


@router.get("/{run_id}/stream")
async def stream_run(
    run_id: str,
    controller: RunController = Depends(get_controller),
):
    """
    SSE stream — yields RunEvents as they occur.
    The Electron UI connects here and updates its local state reactively.

    Event types: state_changed | progress | record_extracted |
                 warning | error | log | checkpoint_saved |
                 run_completed | run_failed | run_cancelled
    """
    try:
        controller.get_run(run_id)  # validate exists
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Run not found: {run_id}")

    return StreamingResponse(
        controller.stream_run(run_id),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "X-Accel-Buffering": "no",  # disable nginx buffering
        },
    )


@router.post("/{run_id}/stop", response_model=RunDetail)
async def stop_run(
    run_id: str,
    controller: RunController = Depends(get_controller),
):
    """Gracefully stop a run. Saves checkpoint before terminating."""
    try:
        state = await controller.stop_run(run_id)
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Run not found: {run_id}")
    return RunDetail(**state.to_dict())


@router.post("/{run_id}/pause", response_model=RunDetail)
async def pause_run(
    run_id: str,
    controller: RunController = Depends(get_controller),
):
    """Pause a running job. Checkpoint is saved; can be resumed."""
    try:
        state = await controller.pause_run(run_id)
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Run not found: {run_id}")
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc))
    return RunDetail(**state.to_dict())


@router.post("/{run_id}/resume", response_model=RunDetail)
async def resume_run(
    run_id: str,
    controller: RunController = Depends(get_controller),
):
    """Resume a paused run from its last checkpoint."""
    try:
        state = await controller.resume_run(run_id)
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Run not found: {run_id}")
    except ValueError as exc:
        raise HTTPException(status_code=409, detail=str(exc))
    return RunDetail(**state.to_dict())


@router.delete("/{run_id}", response_model=RunDetail)
async def cancel_run(
    run_id: str,
    controller: RunController = Depends(get_controller),
):
    """Hard cancel — no checkpoint saved."""
    try:
        state = await controller.cancel_run(run_id)
    except KeyError:
        raise HTTPException(status_code=404, detail=f"Run not found: {run_id}")
    return RunDetail(**state.to_dict())
